core/store/src/db/rocksdb/instance_tracker.rs

```diff
@@ -0,0 +1,171 @@
+use std::sync::{Condvar, Mutex};
+
+use once_cell::sync::Lazy;
+use tracing::info;
+
+/// Describes number of RocksDB instances and sum of their max_open_files.
+///
+/// This is modified by [`InstanceTracker`] which uses RAII to automatically
+/// update the trackers when new RocksDB instances are created and dropped.
+#[derive(Default)]
+struct Inner {
+    /// Number of RocksDB instances open.
+    count: usize,
+
+    /// Sum of max_open_files configuration option of all the RocksDB instances.
+    ///
+    /// This is used when opening multiple instances to guarantee that process
+    /// limit is high enough to allow all of the instances to open the maximum
+    /// number of files allowed in their respective configurations.
+    ///
+    /// For example, if NOFILE limit is set to 1024 and we try to open two
+    /// RocksDB instances with max_open_files set to 512 the second open should
+    /// fail since the limit is too low to accommodate both databases and other
+    /// file descriptor node is opening.
+    max_open_files: u64,
+}
+
+/// Synchronisation wrapper for accessing [`Inner`] singleton.
+#[derive(Default)]
+struct State {
+    mutex: Mutex<Inner>,
+    zero_cvar: Condvar,
+}
+
+impl State {
+    /// Registers a new RocksDB instance and checks if limits are enough for
+    /// given max_open_files configuration option.
+    ///
+    /// Returns an error if process’ resource limits are too low to allow
+    /// max_open_files open file descriptors for the new database instance.
+    fn try_new_instance(&self, max_open_files: u32) -> Result<(), String> {
+        let num_instances = {
+            let mut inner = self.mutex.lock().unwrap();
+            let max_open_files = inner.max_open_files + u64::from(max_open_files);
+            ensure_max_open_files_limit(max_open_files)?;
+            inner.max_open_files = max_open_files;
+            inner.count += 1;
+            inner.count
+        };
+        info!(target: "db", num_instances, "Created a new RocksDB instance.");
+        Ok(())
+    }
+
+    fn drop_instance(&self, max_open_files: u32) {
+        let num_instances = {
+            let mut inner = self.mutex.lock().unwrap();
+            inner.max_open_files = inner.max_open_files.saturating_sub(max_open_files.into());
+            inner.count = inner.count.saturating_sub(1);
+            if inner.count == 0 {
+                self.zero_cvar.notify_all();
+            }
+            inner.count
+        };
+        info!(target: "db", num_instances, "Dropped a RocksDB instance.");
```

Comment:
Same as above logging statement, I think we should capture additional info here.